public/app/containers/Explore/PromQueryField.tsx

```diff
@@ -0,0 +1,310 @@
+            context = 'context-label-values';
+            suggestions.push({
+              label: 'Label values',
+              items: labelValues.map(wrapLabel),
+            });
+          }
+        } else {
+          // Label keys
+          context = 'context-labels';
+          suggestions.push({ label: 'Labels', items: labelKeys.map(wrapLabel) });
+        }
+      } else {
+        refresher = this.fetchMetricLabels(metric);
+      }
+    } else if (_.includes(wrapperClasses, 'context-labels') && !metric) {
+      // Empty name queries
+      const defaultKeys = ['job', 'instance'];
+      // Munge all keys that we have seen together
+      const labelKeys = Object.keys(this.state.labelKeys).reduce((acc, metric) => {
+        return acc.concat(this.state.labelKeys[metric].filter(key => acc.indexOf(key) === -1));
+      }, defaultKeys);
+      if ((text && text.startsWith('=')) || _.includes(wrapperClasses, 'attr-value')) {
+        // Label values
+        if (labelKey) {
+          if (this.state.labelValues[EMPTY_METRIC]) {
+            const labelValues = this.state.labelValues[EMPTY_METRIC][labelKey];
+            context = 'context-label-values';
+            suggestions.push({
+              label: 'Label values',
+              items: labelValues.map(wrapLabel),
+            });
+          } else {
+            // Can only query label values for now (API to query keys is under development)
+            refresher = this.fetchLabelValues(labelKey);
+          }
+        }
+      } else {
+        // Label keys
+        context = 'context-labels';
+        suggestions.push({ label: 'Labels', items: labelKeys.map(wrapLabel) });
+      }
+    } else if (metric && _.includes(wrapperClasses, 'context-aggregation')) {
+      context = 'context-aggregation';
+      const labelKeys = this.state.labelKeys[metric];
+      if (labelKeys) {
+        suggestions.push({ label: 'Labels', items: labelKeys.map(wrapLabel) });
+      } else {
+        refresher = this.fetchMetricLabels(metric);
+      }
+    } else if (
+      // Non-empty but not inside known token unless it's a metric
+      (prefix && !_.includes(wrapperClasses, 'token')) ||
+      prefix === metric ||
+      (prefix === '' && !text.match(/^[)\s]+$/)) || // Empty context or after ')'
+      text.match(/[+\-*/^%]/) // After binary operator
+    ) {
+      suggestions.push({
+        prefixMatch: true,
+        label: 'Functions',
+        items: FUNCTIONS.map(setFunctionMove),
+      });
+
+      if (this.state.metrics) {
+        suggestions.push({
+          label: 'Metrics',
+          items: this.state.metrics.map(wrapLabel),
+        });
+      }
+    }
+
+    return {
+      context,
+      prefix,
+      refresher,
+      suggestions,
+    };
+  }
+
+  request = url => {
+    if (this.props.request) {
+      return this.props.request(url);
+    }
+    return fetch(url);
+  };
+
+  async fetchLabelValues(key) {
+    const url = `/api/v1/label/${key}/values`;
```

Comment:
This part looks similar for next functions:
```js
    const url = `/api/v1/label/${key}/values`;
    try {
      const res = await this.request(url);
      const body = await (res.data || res.json());
```
is it possible to extract it and reuse?